### PR TITLE
#4003: Adding pyind11 to ttnn

### PR DESCRIPTION
**Pull request by eyonland**

### CODE REVIEW

**Review by arakhmati — APPROVED** [review]

**Review by mywoodstock — APPROVED** [review]
LGTM

**Review by tt-aho — APPROVED** [review]

**Review by TT-BrianLiu — APPROVED** [review]

File: ttnn/module.mk [hunk 1]
```diff
@@ -5,3 +5,53 @@
 ttnn/dev_install: python_env/dev
 	echo "Installing editable dev version of ttnn package..."
 	bash -c "source $(PYTHON_ENV)/bin/activate && pip install -e ttnn"
+
+TTNN_LIB = $(LIBDIR)/libttnn.so
```
Comment (tt-rkim) [inline ttnn/module.mk:9]: OK so a couple of things

- Do we plan to use any ttnn C++ sources as a C++ library?
- Do we plan to release separate wheels for ttnn in the future?

For some background: right now we only have the python bindings for eager and the metal runtime as shared objects. Everything else is an archive that is meant to be deleted and not used in production.

**Review by tt-rkim — CHANGES_REQUESTED** [review]

File: ttnn/module.mk (hunk 1 above)
Comment (eyonland) [reply ttnn/module.mk:9]: Yes, we want to have two C++ libraries.  One will be without the pybind11 and is intended for folks who do not want to use any C++.  The other is for pybind11 and is intended for using with python.

Yes, I think we should release separate wheels for ttnn.  Any help on getting this setup would be huge.

File: ttnn/cpp/ttnn/operations/binary.hpp [hunk 2]
```diff
@@ -107,13 +94,8 @@ void py_module(py::module& m_binary) {
...
             return ttnn::reshape(output, original_shape);
-        },
-        py::arg("input_tensor_a"),
-        py::arg("input_tensor_b"),
-        py::kw_only(),
-        py::arg("memory_config") = DRAM_MEMORY_CONFIG);
-}
+        }
 
-}  // namespace binary
```
[the remark this answers is not in the document]
Comment (eyonland) [reply ttnn/cpp/ttnn/operations/binary.hpp:117]: oops

**Review by TT-billteng — CHANGES_REQUESTED** [review]
don't think you need to bump taskflow in this PR

**Review by TT-billteng — APPROVED** [review]

**Review by rayraykay — APPROVED** [review]
Could you please do similar changes to these in a future PR? @nsmithtt made changes so that incremental builds go a bit faster by skipping over trying to install the whole environment again:
https://github.com/tenstorrent-metal/tt-metal/pull/5251/files
https://github.com/tenstorrent-metal/tt-metal/pull/5292/files

**Review by tt-rkim — APPROVED** [review]
Could you please do similar changes to these in a future PR? @nsmithtt made changes so that incremental builds go a bit faster by skipping over trying to install the whole environment again:
https://github.com/tenstorrent-metal/tt-metal/pull/5251/files
https://github.com/tenstorrent-metal/tt-metal/pull/5292/files

File: ttnn/module.mk [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+$(TTNN_PYBIND11_LIB): tt_lib/csrc $(TTNN_PYBIND11_OBJS)
+	@mkdir -p $(LIBDIR)
+	$(CXX) $(TTNN_CFLAGS) $(CXXFLAGS) $(SHARED_LIB_FLAGS) -o $@ $(TTNN_PYBIND11_OBJS) $(TTNN_PYBIND11_LDFLAGS)
+
+
+.PHONY: ttnn/setup_local_so
+ttnn/setup_local_so: $(TTNN_PYBIND11_LIB)
+	rm -f ttnn/ttnn/_ttnn.so
+	cp $^ ttnn/ttnn/_ttnn.so
```
Comment (nsmithtt) [inline ttnn/module.mk:49]: This `.PHONY` target is definitely going to break incremental build.  Can we rewrite this target to be?
```make
ttnn/ttnn/_ttnn.so: $(TTNN_PYBIND11_LIB)
 cp -fp $^ $@
```

Ok if we want to do this as a follow on.

**Review by tt-rkim — APPROVED** [review]

**Pull request merged**